pkg/assets/dynamic/dynamic.go

```diff
@@ -0,0 +1,200 @@
+package dynamic
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/assets/model"
+	"github.com/cozy/cozy-stack/pkg/logger"
+	"github.com/cozy/swift"
+	"github.com/hashicorp/go-multierror"
+)
+
+var ErrDynAssetNotFound = errors.New("Dynamic asset was not found")
+
+var assetsClient = &http.Client{
+	Timeout: 30 * time.Second,
+}
+
+// List dynamic assets
+func ListAssets() (map[string][]*model.Asset, error) {
+	return assetFS.List()
+}
+
+// GetAsset retrieves a raw asset from the dynamic FS and builds a fs.Asset
+func GetAsset(context, name string) (*model.Asset, error) {
+	// Re-constructing the asset struct from the dyn FS content
+	content, err := assetFS.Get(context, name)
+	if err != nil {
+		if err == swift.ObjectNotFound || os.IsNotExist(err) {
+			return nil, ErrDynAssetNotFound
+		}
+		return nil, err
+	}
+
+	h := sha256.New()
+	_, err = h.Write(content)
+	if err != nil {
+		return nil, err
+	}
+	suma := h.Sum(nil)
+	sumx := hex.EncodeToString(suma)
+
+	zippedDataBuf := new(bytes.Buffer)
+	gw := gzip.NewWriter(zippedDataBuf)
+	_, err = gw.Write(content)
+	if err != nil {
+		return nil, err
+	}
+	zippedContent := zippedDataBuf.Bytes()
+
+	asset := model.NewAsset(model.AssetOption{
+		Shasum:   sumx,
+		Name:     name,
+		Context:  context,
+		IsCustom: true,
+	}, zippedContent, content)
+
+	return asset, nil
+}
+
+// RemoveAsset removes a dynamic asset from Swift
+func RemoveAsset(context, name string) error {
+	return assetFS.Remove(context, name)
+}
+
+// RegisterCustomExternals ensures that the assets are in the Swift, and load
+// them from their source if they are not yet available.
+func RegisterCustomExternals(opts []model.AssetOption, maxTryCount int) error {
```

Comment:
I think `RegisterCustomExternals` is always called with a single `AssetOption` in `opts`, and could be simplified to accept a `model.AssetOption` instead of a slice of them.

Reply:
The CLI indeed restricts the dynamic asset insertion to one by one, but the API endpoint gives the option to insert more than one asset in a single request.
We can keep the slice for the API, or change the logic to restrict to a single insertion on the endpoint.